File: daemon/internal/redact/redact.go
package redact

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"regexp"
	"sort"
	"strings"
	"time"
)

const RedactedValue = "***REDACTED***"

var sensitiveKeyParts = []string{
	"API_KEY",
	"SECRET",
	"TOKEN",
	"PASSWORD",
	"CREDENTIAL",
}

var sensitiveTextPattern = regexp.MustCompile(`(?i)\b([A-Z0-9_]*(?:API_KEY|SECRET|TOKEN|PASSWORD|CREDENTIAL)[A-Z0-9_]*)\b(\s*[:=]\s*"?)([^,\s"'\n]+)"?`)

// urlCredentialPattern matches credentials embedded in URLs like postgres://user:pass@host
var urlCredentialPattern = regexp.MustCompile(`(://[^:/@\s]+):([^@\s]+)@`)

type ArtifactMetadata struct {
	CreatedAt time.Time `json:"created_at"`
	ExpiresAt time.Time `json:"expires_at"`
	SHA256    string    `json:"sha256"`
}

func IsSensitiveKeyName(key string) bool {
	upper := strings.ToUpper(key)
	for _, part := range sensitiveKeyParts {
		if strings.Contains(upper, part) {
			return true
		}
	}
	return false
}

func RedactEnviron(environ []string) map[string]string {
	redacted := make(map[string]string, len(environ))
	for _, item := range environ {
		key, value, found := strings.Cut(item, "=")
		key = strings.TrimSpace(key)
		if key == "" {
			continue
		}
		if IsSensitiveKeyName(key) {
			redacted[key] = RedactedValue
		} else if found {
			redacted[key] = RedactText(value)
		} else {
			redacted[key] = ""
		}
	}
	return redacted
}

func RedactText(input string) string {
	result := sensitiveTextPattern.ReplaceAllString(input, "${1}${2}"+RedactedValue)
	result = urlCredentialPattern.ReplaceAllString(result, "${1}:"+RedactedValue+"@")
	return result
}

func BuildMetadata(createdAt time.Time, expiration time.Duration, artifacts map[string][]byte) ArtifactMetadata {
	if expiration <= 0 {
		expiration = 24 * time.Hour
	}
	return ArtifactMetadata{
		CreatedAt: createdAt.UTC(),
		ExpiresAt: createdAt.UTC().Add(expiration),
		SHA256:    ArtifactChecksum(artifacts),
	}
}

func MetadataJSON(createdAt time.Time, expiration time.Duration, artifacts map[string][]byte) ([]byte, error) {
	metadata := BuildMetadata(createdAt, expiration, artifacts)
	return json.MarshalIndent(metadata, "", "  ")
}

func ArtifactChecksum(artifacts map[string][]byte) string {
	names := make([]string, 0, len(artifacts))
	for name := range artifacts {
		names = append(names, name)
	}
	sort.Strings(names)

	hash := sha256.New()
	for _, name := range names {
		_, _ = hash.Write([]byte(name))
		_, _ = hash.Write([]byte{0})
		_, _ = hash.Write(artifacts[name])
		_, _ = hash.Write([]byte{0})
	}

	return hex.EncodeToString(hash.Sum(nil))
}

File: gateway/src/redact.ts
/**
 * Redaction utilities for gateway responses and logs.
 *
 * Mirrors the patterns from daemon/internal/redact/redact.go so that
 * sensitive values (API keys, tokens, passwords, URL credentials) are
 * scrubbed before they leave the gateway in error messages or log output.
 */

export const REDACTED = "***REDACTED***";

const SENSITIVE_KEY_PARTS = [
  "API_KEY",
  "SECRET",
  "TOKEN",
  "PASSWORD",
  "CREDENTIAL",
];

/**
 * Matches inline key=value style sensitive data such as:
 *   MY_API_KEY=abc123
 *   SECRET: "xyzzy"
 */
const sensitiveTextPattern =
  /\b([A-Z0-9_]*(?:API_KEY|SECRET|TOKEN|PASSWORD|CREDENTIAL)[A-Z0-9_]*)\b(\s*[:=]\s*"?)([^,\s"'\n]+)"?/gi;

/**
 * Matches credentials embedded in URLs: scheme://user:password@host
 */
const urlCredentialPattern = /(:\/\/[^:/@\s]+):([^@\s]+)@/g;

/** Returns true if the key name looks like it holds a secret. */
export function isSensitiveKey(key: string): boolean {
  const upper = key.toUpperCase();
  return SENSITIVE_KEY_PARTS.some((part) => upper.includes(part));
}

/**
 * Redact sensitive patterns found in free-form text.
 *
 * - KEY=value / KEY: "value" pairs where KEY contains a sensitive word
 * - URL-embedded credentials (scheme://user:pass@host)
 */
export function redactText(input: string): string {
  // Reset lastIndex for global regexes
  sensitiveTextPattern.lastIndex = 0;
  urlCredentialPattern.lastIndex = 0;

  let result = input.replace(sensitiveTextPattern, `$1$2${REDACTED}`);
  result = result.replace(urlCredentialPattern, `$1:${REDACTED}@`);
  return result;
}

/**
 * Redact an environment-variable style record (key=value strings).
 * Keys whose names look sensitive get their entire value replaced;
 * other values are scanned with {@link redactText}.
 */
export function redactEnviron(environ: string[]): Record<string, string> {
  const out: Record<string, string> = {};
  for (const item of environ) {
    const eqIdx = item.indexOf("=");
    const key = (eqIdx === -1 ? item : item.slice(0, eqIdx)).trim();
    if (!key) continue;
    if (isSensitiveKey(key)) {
      out[key] = REDACTED;
    } else if (eqIdx !== -1) {
      out[key] = redactText(item.slice(eqIdx + 1));
    } else {
      out[key] = "";
    }
  }
  return out;
}

/**
 * Redact sensitive data from a gateway error message string.
 * Convenience wrapper around {@link redactText}.
 */
export function redactErrorMessage(message: string): string {
  return redactText(message);
}
